Source/DafnyBenchmarkingPlugin/JavaBenchmarkCompilationInstrumenter.cs

```diff
@@ -0,0 +1,40 @@
+﻿using Microsoft.Dafny;
+using Microsoft.Dafny.Plugins;
+
+namespace DafnyBenchmarkingPlugin;
+
+public class JavaBenchmarkCompilationInstrumenter : GenericCompilationInstrumenter {
+
+  private readonly ErrorReporter Reporter;
+
+  public JavaBenchmarkCompilationInstrumenter(ErrorReporter reporter) {
+    Reporter = reporter;
+  }
+
+  public override void BeforeClass(TopLevelDecl cls, ConcreteSyntaxTree wr) {
+    if (Attributes.Contains(cls.Attributes, "benchmarks")) {
+      wr.WriteLine("@org.openjdk.jmh.annotations.State(org.openjdk.jmh.annotations.Scope.Benchmark)");
+    }
+  }
+
+  public override void BeforeMethod(Method m, ConcreteSyntaxTree wr) {
+    if (Attributes.Contains(m.EnclosingClass.Attributes, "benchmarks")) {
+      if (m is Constructor) {
+        if (m.Ins.Any()) {
+          Reporter.Error(MessageSource.Compiler, ResolutionErrors.ErrorId.none, m.tok,
+            $"Classes with {{:benchmarks}} can not accept parameters in their constructors");
+        }
+        
+        // _ctor() needs to be explicitly invoked as usual,
+        // and it's convenient (but a bit of a hack) to do this by marking it as Setup.
+        // It's not safe in general to run a Dafny compiled constructor
+        // multiple times on the same object,
+        // so the better solution in the future is probably to maintain the benchmark object
+        // as a separate object that Setup instantiates every time.
```

Comment:
How difficult do you think this would be to implement?

Reply:
Not too complicated, but definitely emitting a fair bit of extra code instead of just some Java annotations. I'll have to do that for C# next anyway (since BenchmarkDotNet doesn't support multithreaded benchmarks directly so I'll have to emit a `Parallel.For` etc.) so I thought it was reasonable to checkpoint this working solution, since in practice the hack is extremely unlikely to cause problems.